### PR TITLE
Add terms prettyer-print

### PR DESCRIPTION
**Pull request by AstroSnail**
Current wips: (yes this is still wip but it might already be an improvement over what's before)

- [x] don't know how to pretty print difficult terms like inferrable.typed and inferrable.application (they are small and can easily lead to confusion when they contain big terms)
- [x] continued invasive changes to terms may be warranted (e.g. closure maybe should store the parameter name?)
- [x] how verbose should it be? e.g. can lambdas/closures be elided whenever they're expected?
- [x] consecutive let terms should be grouped
- [x] chasing a bug that prevents the context from being passed through default-printed terms (i.e. lots of MISSING CONTEXT warnings)
- [ ] need to refactor the test to use actual terms from terms.lua
- [x] need to add a trait that always default-prints terms (literally duplicate prettyprint trait and remove override?)
- [ ] this code is copypaste bonanza, it's extremely not dry

### CODE REVIEW

File: terms.lua [hunk 1]
```diff
@@ -536,11 +735,559 @@ inferrable_term:define_enum("inferrable", {
...
+		end
+
+		pp:_exit()
+	end,
+	typed = function(self, pp, context)
+		local type, usage_counts, typed_term = self:unwrap_typed()
+		context = ensure_context(context)
+
+		pp:_enter()
+		--[[
```
Comment (LunNova) [inline terms.lua:870]: Blockcommented sections could be removed instead?

File: terms.lua [hunk 2]
```diff
@@ -536,11 +735,559 @@ inferrable_term:define_enum("inferrable", {
...
+		pp:unit(pp:_color())
+		pp:unit(") (")
+		pp:unit(pp:_resetcolor())
+
+		pp:any(typed_term, context)
+
+		pp:unit(pp:_color())
+		pp:unit(")")
+		pp:unit(pp:_resetcolor())
+		--[[
```
Comment (LunNova) [inline terms.lua:902]: another commented section

File: term-print.lua
Comment (LunNova) [inline term-print.lua:1]: the words in this file are very very

File: term-print.lua [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+		pp:unit(pp:_resetcolor())
+
+		pp:_indent()
+		pp:_prefix()
+		pp:any(bingle, extra)
+		pp:unit(extra)
+		pp:unit("\n")
+		pp:_dedent()
+
+		--[[
```
Comment (LunNova) [inline term-print.lua:51]: why is quux commented? what is quux? why am i here?

**Comment by AstroSnail** [comment]
@LunNova term-print was the initial sketch file where i discussed with open how the implementation should work. it allowed me to study what techniques would work, and also let me narrow down on bugs

**Comment by AstroSnail** [comment]
it's fine to remove term-print from the pr

**Review by LunNova — APPROVED** [review]

**Pull request merged**